activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

```diff
@@ -639,11 +639,28 @@ def alter_table(
             yield definition if block_given?
           end
 
-          transaction do
-            disable_referential_integrity do
+          fk_enabled = query_value("PRAGMA foreign_keys") == 1
+
+          begin
+            execute("PRAGMA foreign_keys = OFF")
+
+            transaction do
               move_table(table_name, altered_table_name, options.merge(temporary: true))
               move_table(altered_table_name, table_name, &caller)
             end
+
+            execute("PRAGMA foreign_keys = ON") if fk_enabled
+
+            if fk_enabled
+              violations = execute("PRAGMA foreign_key_check").to_a
+              if violations.any?
+                raise ActiveRecord::StatementInvalid,
+                      "Foreign key violations detected after schema change: #{violations.inspect}"
+              end
+            end
```

Comment:
I believe this is supposed to run inside the transaction according the the [sqlite docs](https://sqlite.org/lang_altertable.html) (steps 10-12)

Reply:
Thinking out loud: 

- I wonder if the real problem here is just the ordering of `disable_referential_integrity` and the `transaction`
- `defer_foreign_keys` inside `disable_referential_integrity` seems to be unnecessary as long as it wraps `transaction` and not the opposite
- fixture insertion also has `transaction` wrapping `disable_referential_integrity`, maybe could also be inverted?

Reply:
You're absolutely right, @skipkayhil, regarding the order - it's fixed now.

Regarding fixtures insertion, should we address this in the same PR or create a separate issue/PR for the fixture loading bug?